### ISSUE TITLE
Add JSON function support in DuckDB

### ISSUE THREAD

**Issue by mathieusanchez** [jOOQ/jOOQ#17074]
### Use case

We want to replace BigQuery for DuckDB on small clients.

We need the JSON functions support for DuckDB.
JSON functions we use:
* `jsonEntry`
* `jsonObject`
* `jsonArray`
* `jsonObjectAgg`
* `jsonArrayAgg`

### Possible solution

https://duckdb.org/docs/extensions/json.html

### jOOQ Version

jOOQ Enterprise Edition 3.20-SNAPSHOT

### Database product and version

DuckDB 1.0.0

### Java Version

openJDK 17.0.8.1 2023-08-24

**Comment by lukaseder** [comment]
Thanks for your suggestion.

I don't recall why we haven't added support yet. There's a brief bullet about adding support for the `JSON` type in this list:

- https://github.com/jOOQ/jOOQ/issues/15732

But no `JSON` function support. Will look into this right away.

**Comment by lukaseder** [comment]
Hmm, I distinctly recall having rejected using the JSON support for `MULTISET` emulations, because the `JSON_GROUP_ARRAY` function cannot specify `ORDER` or `FILTER`

**Comment by lukaseder** [comment]
... but then again, this is possible, so we might be able to work around limitations:

```sql
SELECT to_json(array_agg(DISTINCT json_object('v', v) ORDER BY v desc) FILTER (WHERE v != 3))
FROM (VALUES (1),(2),(2),(3)) AS t (v)
```

**Comment by lukaseder** [comment]
Here's the `JSON_OBJECTAGG` version:

```sql
SELECT to_json(array_agg(MAP([k], [v])) FILTER (WHERE k != 'c'))
FROM (VALUES ('a', 1),('b', 2),('c', 3)) AS t (k, v);
```

**Comment by lukaseder** [comment]
Some DuckDB bugs discovered during investigations:

- https://github.com/duckdb/duckdb/issues/13436
- https://github.com/duckdb/duckdb/issues/13437

Specifically, it's not possible currently, to distinguish between `CAST('null' AS JSON)` and `CAST(NULL AS JSON)` in certain contexts.

**Comment by lukaseder** [comment]
These things will be supported:

- [x] `JSON_ARRAY`
  - [x] `ABSENT ON NULL` via `ARRAY` and `ARRAY_FILTER`
- [x] `JSON_OBJECT`
  - [x] `ABSENT ON NULL` via `MAP_OF_ENTRIES` and `ARRAY_FILTER`
- [x] `JSON_VALUE`
- [x] `JSON_EXISTS`
- [x] `JSON_ARRAY_LENGTH`
- [x] `JSON_KEY_EXISTS` (via `JSON_GET_ATTRIBUTE`)
- [x] `JSON_GET_ATTRIBUTE` (as text, too)
- [x] `JSON_GET_ELEMENT` (as text, too)
- [x] `JSON_ARRAYAGG` (via `TO_JSON(ARRAY_AGG(..))`)
  - [x] `DISTINCT` is supported natively in `ARRAY_AGG`
  - [x] `ORDER BY` is supported natively in `ARRAY_AGG`
  - [x] `ABSENT ON NULL` via `ARRAY_FILTER`
  - [x] `OVER` is supported natively in `ARRAY_AGG`
  - [x] `FILTER` emulation
    - [x] `ABSENT ON NULL` via `ARRAY_FILTER`
    - [x] `NULL ON NULL`
- [x] `JSON_OBJECTAGG` (via `TO_JSON(MAP(ARRAY_AGG(..), ARRAY_AGG(..)))` or similar)

These won't be supported as there's no equivalent functionality in DuckDB, I think.

- `JSON_INSERT`
- `JSON_REMOVE`
- `JSON_REPLACE`
- `JSON_SET`
- `JSON_TABLE`

**Comment by lukaseder** [comment]
Implemented in 3.20.0
